src/examples/libpmemobj/libart/libart2/art.c

```diff
@@ -0,0 +1,1654 @@
+static void remove_child48(PMEMobjpool *pop, PMEMoid n,
+			    PMEMoid *ref, unsigned char c);
+static void remove_child256(PMEMobjpool *pop, PMEMoid n,
+			    PMEMoid *ref, unsigned char c);
+
+static PMEMoid *find_child(PMEMoid n, uint64_t type_num, unsigned char c);
+
+static int check_prefix(const art_node *n, const unsigned char *key,
+			    int key_len, int depth);
+
+PMEMoid art_minimum(TOID(struct art_tree_root) t);
+PMEMoid art_maximum(TOID(struct art_tree_root) t);
+
+static void destroy_node(PMEMoid n);
+int art_iter(PMEMobjpool *pop, art_callback cb, void *data);
+int art_iter2(PMEMobjpool *pop, art_callback cb, void *data);
+
+static void PMEMOIDcopy(PMEMoid *dest, const PMEMoid *src, const int n);
+static void PMEMOIDmove(PMEMoid *dest, PMEMoid *src, const int n);
+
+static void
+PMEMOIDcopy(PMEMoid *dest, const PMEMoid *src, const int n)
+{
+	int i;
+
+	for (i = 0; i < n; i++) {
+		dest[i] = src[i];
+	}
+}
+
+static void
+PMEMOIDmove(PMEMoid *dest, PMEMoid *src, const int n)
+{
+	int i;
+
+	if (dest > src) {
+		for (i = n - 1; i >= 0; i--) {
+			dest[i] = src[i];
+		}
+	} else {
+		for (i = 0; i < n; i++) {
+			dest[i] = src[i];
+		}
+	}
+}
+
+PMEMoid
+alloc_node(PMEMobjpool *pop, art_node_type node_type, int buffer_size)
+{
+	PMEMoid an = null_oid;
+	art_leaf *alp;
+
+	switch (node_type) {
+	case NODE4:
+		an = pmemobj_tx_zalloc(sizeof(art_node4),
+				    art_node4_type_num);
+		break;
+	case NODE16:
+		an = pmemobj_tx_zalloc(sizeof(art_node16),
+				    art_node16_type_num);
+		break;
+	case NODE48:
+		an = pmemobj_tx_zalloc(sizeof(art_node48),
+				    art_node48_type_num);
+		break;
+	case NODE256:
+		an = pmemobj_tx_zalloc(sizeof(art_node256),
+				    art_node256_type_num);
+		break;
+	case art_leaf_t:
+		an = pmemobj_tx_zalloc(sizeof(art_leaf) + buffer_size,
+				    art_leaf_type_num);
+		alp = pmemobj_direct(an);
+		alp->buffer_len = buffer_size;
+		break;
+	default:
+		/* invalid node type */
+		break;
+	}
+
+	return an;
+}
+
+int
+art_tree_init(PMEMobjpool *pop, int *newpool)
+{
+	int errors = 0;
+	TOID(struct art_tree_root) root;
+
+	if (pop == NULL) {
+		errors++;
+	}
+
+	null_oid = OID_NULL;
+
+	art_tree_root_type_num = TOID_TYPE_NUM(struct art_tree_root);
+	art_node4_type_num = TOID_TYPE_NUM(art_node4);
+	art_node16_type_num = TOID_TYPE_NUM(art_node16);
+	art_node48_type_num = TOID_TYPE_NUM(art_node48);
+	art_node256_type_num = TOID_TYPE_NUM(art_node256);
+	art_leaf_type_num = TOID_TYPE_NUM(art_leaf);
+
+	if (!errors) {
+		TX_BEGIN(pop) {
+			root = POBJ_ROOT(pop, struct art_tree_root);
+			if (*newpool) {
+				TX_ADD(root);
+				D_RW(root)->root = OID_NULL;
+				D_RW(root)->size = 0;
+				*newpool = 0;
+			}
+		} TX_END
+	}
+
+	return errors;
+}
+
+// Recursively destroys the tree
+static void
+destroy_node(PMEMoid n)
+{
+	uint64_t type_num;
+	art_node4 *an4;
+	art_node16 *an16;
+	art_node48 *an48;
+	art_node256 *an256;
+
+	// Break if null
+	if (OID_IS_NULL(n))
+		return;
+
+	type_num = pmemobj_type_num(n);
```

Comment:
There are many places where I use pmemobj_type_num() to do explicit type checking in libart2. What I wanted to avoid in libart2 was primarily related to TOID*. Using TOID* created some problems in the initial libart port as in art, but also in other trees, references within the tree are of generic nature and I think that TOID* is not very well suited to implement generic references.